File: data structures and algorithms/cpp/08_031.cpp
#include <unordered_set>
#include <vector>

bool hasNearbyDuplicates(std::vector<int>& nums, int k) {
    std::unordered_set<int> window;
    
    for (int i = 0; i < nums.size(); i++) {
        if (i >= k) {
            window.erase(nums[i - k]);
        }
        
        if (window.count(nums[i]) > 0) {
            return true;
        }
        
        window.insert(nums[i]);
    }
    
    return false;
}

File: python/08_031.py
def has_nearby_duplicates(nums, k):

    window = set()
    
    for i, num in enumerate(nums):
        # Remove element outside window
        if i >= k:
            window.remove(nums[i - k])
            
        # Check if current element is in window
        if num in window:
            return True
            
        window.add(num)
    
    return False